docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,280 @@
+replacing 4.12.0 with 4.12.1) to make sure we don't grow too much our configuration
+files. As per special requests (usually made by customers), we append additional
+versions. For example, see:
+https://gitlab.cee.redhat.com/service/app-interface/-/blob/3fb1ac40a81c7de9bdf1ae6262798fee9336debf/data/services/assisted-installer/cicd/target/production/assisted-service.yaml
+
+The idea of this enhancement is configuring the assisted-service to the OpenShift
+channels it needs to follow (e.g. ``candidate-4.12``, ``stable-4.11``), and let
+it periodically fetch versions to provide them to customers.
+
+## Motivation
+
+It becomes a pretty tedious process to handle each version separately. Some
+use-cases we tend to have:
+* A certain customer use a specific version x.y.z for certification, and we replace
+  this version with a newer one x.y.z+w (w>=1). In this case we'll append again the
+  x.y.z version to the configuration.
+* As described above, we do version updates either automatically or manually.
+* QE team wants to check if a newly added OpenShift version has a regression in its
+  behavior from a previous version.
+* QE team tests a flow which includes OpenShift upgrade process.
+
+### Goals
+
+* Assisted-service is able to provide all versions included in the upgrade-channels,
+  based on provided configuration.
+* We keep ourselves the option to specifically configure certain versions as we can
+  nowadays. This is to make sure we can handle edge-cases and be flexible if needed.
+* UI needs to be adjusted to handle this growing list of versions (at the time being,
+  a total of ~140 versions).
+
+### Non-Goals
+
+* Dynamic handling of y-versions, live-iso discovery images, and such. A new minor
+  OpenShift version is introduced every 4 months, requires additional code changes
+  (for example in feature-support-level list) and it isn't worth the costs and
+  implications of making assisted-image-service refresh itself and pull-in new
+  images.
+* Handling OpenShift versions in a kubernetes API interface. For this use-case
+  versions are externally configured in the ``ClusterImageSet`` custom resource.
+
+## Proposal
+
+The assisted-service will make use of a new configuration variable called
+``DYNAMIC_OPENSHIFT_RELEASES_CONFIG``, which outlines the supported matrix. For
+example:
+
+```
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='{
+   "4.11": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["stable"]
+         }
+      ]
+   },
+   "4.12": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "arm64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "multi",
+            "channels": ["candidate", "stable"]
+         }
+      ]
+   }
+}'
+```
```

Comment:
Do we need some API to fetch this list? Maybe so the user could ensure it's configured properly.

Reply:
nice idea. added